docs/language-specification.md

```diff
@@ -0,0 +1,94 @@
+or combined conjunctively using `&&` or disjunctively using `||`, allowing
+
+
+## Grammar
+```
+// Constraints are expressions of type bool.
+expression ::=
+  | 'true' | 'false'                                               // Boolean constants.
+  | numeral                                                        // Numeric constants.
+  | key                                                            // Table keys.
+  | '!' expression                                                 // Boolean negation.
+  | '-' expression                                                 // Arithmetic negation.
+  | '(' expression ')'                                             // Parentheses.
+  | expression '::' id                                             // Field access (projection).
+  | expression ('&&' | '||' | '->' | ';') expression               // Binary boolean operators.
+  | expression ('==' | '!=' | '>' | '>=' | '<' | '<=') expression  // Comparisons.
+
+key ::= id ('.' id)*                                               // Table keys, e.g. "hdr.ethernet.eth_type".
+id ::= [_a-zA-Z][_a-zA-Z0-9]*                                      // Identifiers.
+
+numeral ::=
+  | (0[dD])? [0-9]+                                                // Decimal numerals.
+  | 0[bB] [0-1]+                                                   // Binary numerals.
+  | 0[oO] [0-7]+                                                   // Octary numerals.
+  | 0[xX] [0-9a-fA-F]+                                             // Hexadecimal numerals.
+```
+Top-level expression may optionally be terminated by a trailing ';'.
+
+### Precedence and Associativity
+
+Operators ordered by precedence: higher in the table means "binds stronger".
+
+| Syntax               | Semantics           | Associativity | Examples                                                                 |
+|----------------------|---------------------|---------------|--------------------------------------------------------------------------|
+| ::                   | Field access        | N/A           | ipv4.dst::prefix_length                                                  |
+| !                    | Boolean negation    | N/A           | !true, !(key == 2), !!false                                              |
+| -                    | Arithmetic negation | N/A           | -1, -(10 + 3), --2                                                       |
+| ==, !=, >, >=, <, <= | Comparison          | none          | 2 == 4, 1 > -12, true != false                                           |
+| &&                   | Boolean conjunction | left          | ipv4.valid == 1 && ipv4.dst == 0xf0f0f0f0                                |
+| \|\|                 | Boolean disjunction | left          | ipv4.valid == 1 \|\| ipv6.valid == 1                                     |
+| ->                   | Boolean implication | none          | ipv4.dst::mask != 0 -> ethernet.ether_type == 0x800                      |
```

Comment:
Currently `->` is right-associative. Make it none-associative (meaning the programmer will have to disambiguate using parentheses) instead, since associativity actually affects the semantics and the casual programmer will likely not expect `->` to be right-assoicative.